File: frontend/src/types/telemetry.ts
export type DeviceStatus = "online" | "offline" | "error";
export type DeviceMode = "normal" | "maintenance" | "safe";

export type DeviceError =
    | "sensor_failure"
    | "overheat"
    | "communication_timeout"
    | "voltage_drop"
    | null;

export interface TelemetryPayload {
    status: DeviceStatus;
    mode: DeviceMode;
    temperature: number;
    pressure: number;
    voltage: number;
    rpm: number;
    connection_quality: number;
    error: DeviceError;
    timestamp: string;
}

export interface WebSocketMessage {
    type: "telemetry";
    payload: TelemetryPayload;
}

File: backend/app/schemas/telemetry.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.domain.enums import DeviceError


class TelemetryResponse(BaseModel):
    temperature: float = Field(..., ge=-100, le=300)
    pressure: float = Field(..., ge=0, le=20)
    voltage: float = Field(..., ge=0, le=100)
    rpm: int = Field(..., ge=0, le=10000)
    connection_quality: int = Field(..., ge=0, le=100)
    error: DeviceError | None = None
    timestamp: datetime